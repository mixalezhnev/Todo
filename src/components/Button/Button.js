import React, { Component } from 'react';
import '../../style/Button.css';

class Button extends Component {
  handleClick(e) {
    e.preventDefault();
    alert('Oh yea ;)');
  }

  render(props) {
    return (
      <a href="#" className="Button" onClick={this.handleClick}> 
        { this.props.text }
      </a>
    )
  }
}

export default Button;