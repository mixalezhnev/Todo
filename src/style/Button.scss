.Button {
  width: 100px;
  min-height: 36px;
  border: 0;
  outline: none;
  text-align: center;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  padding: 0 11px;
  background-color: #1B89DE;
  border-radius: 4px;
  font-family: sans-serif;
  font-weight: 600;
  text-decoration: none;
  color: white;
  margin-bottom: 5px;
}

.Button-red {
  background: #FB8888;
}

.Button-grey {
  background: #5E6472;
}