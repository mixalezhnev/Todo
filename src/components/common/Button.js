import React from 'react';

import './Button.res/style.css'

const Button = (props) => {
  const {type, styles} = props;
  
  return  <button style={styles} className={`Button Button-${type}`}></button>;
}

Button.defaultProps = {
  type: 'add',
  styles: { 
    top: 0,
    left: 0
  }
}

export default Button;
