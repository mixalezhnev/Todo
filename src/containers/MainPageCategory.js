import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import {
  addNestedCategory,
  deleteCategory,
  renameCategory
} from '../actions/CategoryActions';
import { getTasks } from '../actions/TaskListActions';

import Button from '../components/Button';
import Category from '../components/Category';
import ShareExtendState from './ShareExtendState';

import './CategoryContainer.res/style.css';

export class Container extends Component {
  static propTypes = {
    id: PropTypes.node.isRequired,
    title: PropTypes.string.isRequired,
    renameCategory: PropTypes.func.isRequired,
    addNestedCategory: PropTypes.func.isRequired,
    deleteCategory: PropTypes.func.isRequired,
    getTasks: PropTypes.func.isRequired,
    extended: PropTypes.bool.isRequired,
    onExtend: PropTypes.func.isRequired,
    getChildrenCount: PropTypes.func.isRequired,
    currentCategoryId: PropTypes.string.isRequired,
    children: PropTypes.oneOfType(
      [
        PropTypes.arrayOf(PropTypes.element.isRequired),
        PropTypes.bool,
        PropTypes.object
      ]
    )
  }

  handleRenameCategoryClick = () => {
    const {
      id,
      title,
      renameCategory
    } = this.props;
    const newTitle = prompt("Enter new title", title);
    if (!newTitle) return;

    renameCategory(id, newTitle);
  }

  handleAddNestedCategoryClick = () => {
    const {
      id,
      addNestedCategory
    } = this.props;
    const title = prompt("Enter title");
    if (!title) return;

    addNestedCategory(id, title);
  }

  handleDeleteCategoryClick = () => {
    const {
      id,
      deleteCategory,
      currentCategoryId,
      getTasks,
    } = this.props;

    if (!confirm("Are you sure about this?")) return;

    deleteCategory(id);
    if (currentCategoryId === id) {
      getTasks(null);
    }
  }

  render() {
    const {
      id,
      title,
      extended,
      onExtend,
      getChildrenCount,
      children
    } = this.props;

    return (
      <div className='my-categoryContainer-component'>
        <Category id={id} title={title} onExtend={getChildrenCount(children) > 0 && onExtend}>
          <Button
            type="edit"
            onClick={this.handleRenameCategoryClick}
            style={{ marginLeft: '10px', marginRight: 'auto'}}/>
          <Button
            type="delete"
            onClick={this.handleDeleteCategoryClick}/>
          <Button
            type="add"
            onClick={this.handleAddNestedCategoryClick}/>
        </Category>
        {children && 
          <div className={extended
            ? "categories categories_extended"
            : "categories categories_hidden"}>{children}</div>}
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addNestedCategory: bindActionCreators(addNestedCategory, dispatch),
    deleteCategory: bindActionCreators(deleteCategory, dispatch),
    renameCategory: bindActionCreators(renameCategory, dispatch),
    getTasks: bindActionCreators(getTasks, dispatch),
  }
}

const MainPageCategoryContainer = ShareExtendState(Container);
export default connect(null, mapDispatchToProps)(MainPageCategoryContainer);