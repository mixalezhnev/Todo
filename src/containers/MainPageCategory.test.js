import React from 'react';
import { shallow, mount } from 'enzyme';

import { Container } from './MainPageCategory';
import Button from '../components/Button';

const setup = () => {
  const props = {
    id: 0,
    title: 'New title',
    renameCategory: jest.fn(),
    addNestedCategory: jest.fn(),
    deleteCategory: jest.fn(),
    getTasks: jest.fn(),
    extended: true,
    onExtend: () => {},
    currentCategoryId: 0,
    getChildrenCount: jest.fn(() => 1)
  };

  const wrapper = shallow(
    <Container {...props}>
      <div className="testClass"></div>
    </Container>
  );

  return {
    wrapper,
    props
  }
};

describe('Containers :: MainPageCategory', () => {
  it('should render self and subcomponents', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render self and subcomponents without children', () => {
    const { props } = setup();
    const newProps = {
      ...props,
      getChildrenCount: jest.fn(() => 0)
    }
    const wrapper = shallow(<Container {...newProps}></Container>);

    expect(wrapper).toMatchSnapshot();
  });

  it(`should render self and subcomponents with 
    not extended container`, () => {
    const { props } = setup();
    const newProps = {
      ...props,
      extended: false
    }
    const wrapper = shallow(
      <Container {...newProps}>
        <div className="testClass"></div>
      </Container>
      );

    expect(wrapper).toMatchSnapshot();
  });

  it('should call renameCategory', () => {
    const { wrapper, props } = setup();
    const newTitle = 'new title';

    window.prompt = jest.fn(() => newTitle);
    wrapper.find(Button).at(0).prop('onClick')();

    expect(props.renameCategory)
      .toBeCalledWith(props.id, newTitle);
  });
  
  it('should prevent renameCategory call', () => {
    const { wrapper, props } = setup();

    window.prompt = jest.fn(() => '');
    wrapper.find(Button).at(0).prop('onClick')();

    expect(props.renameCategory)
      .not
      .toBeCalled();
  });
  
  it('should call addNestedCategory', () => {
    const { wrapper, props } = setup();
    const newTitle = 'new title';

    window.prompt = jest.fn(() => newTitle);
    wrapper.find(Button).at(2).prop('onClick')();

    expect(props.addNestedCategory)
      .toBeCalledWith(props.id, newTitle);
  });

  it('should prevent addNestedCategory call', () => {
    const { wrapper, props } = setup();

    window.prompt = jest.fn(() => '');
    wrapper.find(Button).at(2).prop('onClick')();

    expect(props.addNestedCategory)
      .not
      .toBeCalled();
  });
  
  it('should call deleteCategory and getTasks', () => {
    const { wrapper, props } = setup();

    window.confirm = jest.fn(() => true);
    wrapper.find(Button).at(1).prop('onClick')();

    expect(props.deleteCategory)
      .toBeCalledWith(props.id);
    expect(props.getTasks)
      .toBeCalledWith(null);
  });
  
  it('should prevent deleteCategory call', () => {
    const { wrapper, props } = setup();

    window.confirm = jest.fn(() => false);
    wrapper.find(Button).at(1).prop('onClick')();

    expect(props.deleteCategory)
      .not
      .toBeCalled();
  });
  
  it('should prevent getTasks call', () => {
    const { props } = setup();
    const newProps = {
      ...props,
      id: 41
    };
    const wrapper = shallow(<Container {...newProps}></Container>);

    window.confirm = jest.fn(() => true);
    wrapper.find(Button).at(1).prop('onClick')();

    expect(props.getTasks)
      .toBeCalledWith(newProps.id);
  });
});
