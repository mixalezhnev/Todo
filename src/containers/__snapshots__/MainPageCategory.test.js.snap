exports[`containers MainPageCategory should render self and subcomponents 1`] = `
<div
  className="my-categoryContainer-component">
  <Category
    id="0"
    onExtend={[Function]}
    title="New title">
    <Button
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "10px",
          "marginRight": "auto",
        }
      }
      type="edit" />
    <Button
      onClick={[Function]}
      type="delete" />
    <Button
      onClick={[Function]}
      type="add" />
  </Category>
  <div
    className="categories categories_extended">
    <div
      className="testClass" />
  </div>
</div>
`;

exports[`containers MainPageCategory should render self and subcomponents with 
      not extended container 1`] = `
<div
  className="my-categoryContainer-component">
  <Category
    id="0"
    onExtend={[Function]}
    title="New title">
    <Button
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "10px",
          "marginRight": "auto",
        }
      }
      type="edit" />
    <Button
      onClick={[Function]}
      type="delete" />
    <Button
      onClick={[Function]}
      type="add" />
  </Category>
  <div
    className="categories categories_hidden">
    <div
      className="testClass" />
  </div>
</div>
`;

exports[`containers MainPageCategory should render self and subcomponents without children 1`] = `
<div
  className="my-categoryContainer-component">
  <Category
    id="0"
    onExtend={false}
    title="New title">
    <Button
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "10px",
          "marginRight": "auto",
        }
      }
      type="edit" />
    <Button
      onClick={[Function]}
      type="delete" />
    <Button
      onClick={[Function]}
      type="add" />
  </Category>
</div>
`;
